Upgrade to support higher security hashes and deriving encrypt AND signing keys

grempe opened the pull request:
Hi, I needed pretty much what you had built (thanks) but I also wanted to generate high quality signing keys as well as encryption keys.

I have gone through the code, tests, and readme and made significant updates including the concept of `v1` and `v2` versions of the main function.  v1 is fully backward compatible and v2 provides higher security through as 64 byte BLAKE2b (vs. BLAKE2s) keyed hash and uses 64 bytes of scrypt output.  The 64 bytes of output is split in two to provide two different seeds, one for generating encryption keys, and one for signing keys.

I also return the Base64 version of the keys for convenience. The tests vectors for v1are unmodified from your original and I've added new tests for the v2 version which is now the default. A user upgrade from the old version should only require adding the 'v1' version argument to their existing code block.

I bumped the version of the package to 2.0.0 to fit with semver.

I hope you can pull this changeset upstream and push a new version.

Thanks!

Glenn

grempe committed:
Use current release versions of all dependencies

- Avoid use of tweetnacl fork
- Update tweetnacl to current release version 0.14.0
- Incorporate tweetnacl-util which was extracted from tweetnacl recently
- Update scrypt-async to 1.2.0
- npm build dist/session25519.js with updates

grempe committed:
Minor stylistic cleanup. http://standardjs.com compliant

grempe committed:
Run node tests against more current releases as well

grempe committed:
README enhancements

grempe committed:
standard.js catch {} should be on its own line

grempe committed:
Optimize. Eliminate some string conversions (scrypt bytes -> base64 -> Uint8Array)

grempe committed:
Comments and whitespace

grempe committed:
Eliminate two variable assignments

grempe committed:
Consistent in use of return

grempe committed:
The comment URL on a single line

grempe committed:
Generate ed25519 signing key pairs. README enhancements.

grempe committed:
Rebuild dist/session25519.js

grempe committed:
browserify --standalone build

grempe committed:
Use blakejs in place of blake2s-js, byte compatible and opens up chance to use 64 bit BLAKE2b

grempe committed:
Support higher security BLAKE2b 64 Byte hash. Derive encrypt and signing keys. Backwards compatible

grempe committed:
Use Uint8Array.subarray() instead of Uint8Array.slice() which has less browser support.

grempe committed:
Typo

grempe committed:
Added two packages that were inadvertantly left out of README

jo commented:
Thank you @grempe for this epic and very much appreciated pull request! I have read through your changes carefully and absolutely agree with both the idea and your implementation. Much kudos for your documention improvements, or better say for making a documentation at all, since before there was barely.

Please give me one more night to think about the implications and your changes. There is so much to learn for me in there.

My first question is not really related to the merge of this pull: While reading your commits one by one I stumbled upon your [idea of using a hash of the private key for the signing key](https://github.com/jo/session25519/pull/2/commits/429bd628a8ef9e48768b9c56bcc2eacf014d5540#diff-04c6e90faac2675aa89e2176d2eec7d8R74) and immediately thought this might be insecure. And you also replaced that method by [generating a 64-bit private key and split it into two 32-bit private key seeds](https://github.com/jo/session25519/pull/2/commits/b7c96982fef4ff0cecc753f523bf81ec0ff92137#diff-04c6e90faac2675aa89e2176d2eec7d8R145) later. But why does this trigger my alarm bells? Would it be secure?

The second thing I found is that the size of the build `dist/session25519.js` has increased from   107 to 143 KB ([2,469 additions, 1,457 deletions](https://github.com/jo/session25519/pull/2/files#diff-b019c02c4170fe29fab1e7e6e1b764c3)). Not that its a problem just for me to know - is this because of [the replacement of `blake2s-js` with `blakejs`](https://github.com/jo/session25519/pull/2/commits/5c2223231a60c5b2b5419261a3f9f9ae1950d2cd)?

Thanks!

jo commented on README.md:
```diff
@@ -1,43 +1,201 @@
+
+The new `v2` version enhances the security of the system, and provides an
+additional key pair. The hash of the password is now a 64 Byte (vs. 32B in v1)
+`BLAKE2b` hash (vs. `BLAKE2s` in `v1`). This hash is alse keyed, like an HMAC,
```
a typo?

``` diff
- This hash is alse keyed
+ This hash is also keyed
```

grempe replied:
This line was removed in updated README. Good catch though!

jo commented on README.md:
````diff
@@ -1,43 +1,201 @@
+
 ```sh
-npm test
+npm run test
````
Why this change?

grempe replied:
Fixed in latest commits.

grempe commented:
Hi @jo, glad you are enjoying the pull request.  To answer your main question about the security of splitting the 64 Byte hash in two in order to use the two halves as seeds. I was implementing something along these lines a few months ago and I asked @dchest (author of tweetnacl.js) about the security of this and he responded in an email.

```
If your user has 32 secret random bytes, but you need 64 random bytes — 32 bytes for nacl.box secret key, and 32 bytes for nacl.sign secret key seed — the correct answer is not to look into the guts of algorithms to figure out if their relationship is safe, but to cryptographically expand those 32 bytes into 64 bytes. If you’re not planning to derive more keys from it in the future, a simple SHA512 will work fine:

  var secret = nacl.randomBytes(32); // this is what a user will store
  var expandedSecret = nacl.hash(secret);
  var boxKeyPair = nacl.box.keyPair.fromSecretKey(expandedSecret.subarray(0, 32));
  var signKeyPair = nacl.sign.keyPair.fromSeed(expandedSecret.subarray(32, 64)); 

By the way, you can have a single _keypair_ for both signing and boxing, by calculating a DH keypair from the signing one: see https://github.com/dchest/ed2curve-js (There’s no proof that this is safe yet, but it is considered safe by cryptographers.)
```

While writing this he just weighed in on Twitter saying the same:

https://twitter.com/dchest/status/711597559563862016

grempe committed:
gitignore output from failed test runs

grempe committed:
No need for different hash inputs to scrypt. Simplify.

Based on feedback from @dchest in a Twitter conversation starting
here:

https://twitter.com/dchest/status/711596208414969856

Now the only diff between v1 and v2 input to scrypt is the dkLen
which will be 32B for v1 and 64B for v2.

Both use BLAKE2s pre-hash.

grempe committed:
The 'run' flag is not needed for 'npm test', only for build.

grempe committed:
Updated README to reflect recent changes

grempe commented:
@jo here are some new commits on the pull request that reflect your questions and the discussion on Twitter today with @dchest

The code is now a bit simpler, using BLAKE2s 32B hash as the scrypt input for both v1 and v2 as @dchest thought the use of the 64 Byte hash didn't really offer that much security wise.

Regarding your file size question I suspect the increase is due to both the change in blake lib and in the use of the browserify --standalone flag which I needed to make it work in the browser for me.  I wasn't watching it too closely though.

I had previously bumped the version string up to 2.0, I'll leave it up to you if that is ok or if you want to use 2.1 for these new commits even though 2.0 was never released.

I added a shout-out to @dchest in the README.md as well. He deserves it. :-)

Cheers

grempe committed:
Extract the single function used from tweetnacl-util-js and remove that extra lib.

grempe commented:
Oh, and another reason for the small size increase in the dist file is the addition of the base64 lib.

grempe commented:
I removed the dependency on the recently extracted nacl-util-js package.  I had reduced its usage to only a single call and I extracted the code from @dchest code and embedded it here to simplify.

Question. You have the .travis.yml file (which I updated) but you don't seem to be running it automatically on builds. If you are using that it would be good to push this through CI to make sure things are OK.

jo commented:
I activated CI on Travis, thanks for pointing that out!

grempe committed:
On the path to removing v1/v2. Intermediate step

grempe committed:
Greatly simplified. Removed v1 and v2 version scheme.

This was possible due to the guidance of @dchest and the
realization that the first 32 Bytes of scrypt output will
be identical when dkLen is specified as 32 or 64.

This means we can just *always* generate 64 Bytes of derived
key material and previous username/password combininations
will still generate the same exact keys. All users
now just get the extra signing keys and Base64 keys for free.

Removing code, especially in a security tool, is a good thing.

Good collaboration.

grempe committed:
Add compiler for Travis CI to try and fix build issue.

grempe committed:
Ugh, add base64-js depend in package.json

grempe committed:
Revert back to @dchest blake2s-js now that BLAKE2b no longer needed. Should pass on Node < v4

grempe commented:
ok, all green now.  Unless you see other issues or enhancements I think it is good to go.  Good collaboration with @jo and @dchest.

jo commented:
Good work! Will do final review tomorrow.

\o/

jo commented on .travis.yml:
```diff
@@ -1,6 +1,20 @@
+  - "0.10"
   - "iojs"
+env:
+  - CXX=g++-4.8
```
Is this still needed?

jo commented on package.json:
```diff
@@ -1,11 +1,11 @@
   "name": "session25519",
-  "version": "1.0.0",
-  "description": "Derive curve25519 key pair from username/password via scrypt.",
+  "version": "2.0.0",
```
No need for breaking release, this is definitively `1.1.0` :100:

jo commented on README.md:
```diff
@@ -1,43 +1,161 @@
+[zxcvbn](https://github.com/dropbox/zxcvbn) are critically important to
+the overall security of the system.
+
+## Version 2.x.x Changes
```
``` diff
- ## Version 2.x.x Changes 
+ ## Version 1.1.x Changes 
```

jo commented:
I made some comments, everything else looks very good to me.

Very good collaboration indeed!

grempe committed:
Remove probably unneeded experimental Travis build config.

grempe committed:
No longer breaking changes. So semver version is 1.1.x

grempe commented:
Changes made.  All green.

jo commented:
Awesome!

jo commented:
```
[jo@linux session25519]$ npm publish
+ session25519@1.1.0
```

elimisteve commented on test.js:
```diff
@@ -1,17 +1,24 @@
-      secretKey: nacl.util.decodeBase64('p6FU22u0423AcS0AKY6I+XeBTMDzQLX57XOH9Xg28pw=')
-    }, 'correct key has been calculated')
-    
+      publicKey: base64.toByteArray('EHfJ2sJJt7LjamnuJul6TlCb5SkgAwJI1PP7q95jh04='),
```
Same as before, yay! Therefore this appears backward-compatible with v1, and therefore theoretically compatible with miniLock, right @grempe?

elimisteve replied:
Ah, one just needs to take the `Uint8Array` representation of the public key and then base58-encode it to get the "miniLock ID", it appears, but the derived key pair should be identical, methinks. /cc @jo

elimisteve commented on test.js:
```diff
@@ -1,17 +1,24 @@
-    
+      publicKey: base64.toByteArray('EHfJ2sJJt7LjamnuJul6TlCb5SkgAwJI1PP7q95jh04='),
+      publicKeyBase64: 'EHfJ2sJJt7LjamnuJul6TlCb5SkgAwJI1PP7q95jh04=',
+      secretKey: base64.toByteArray('p6FU22u0423AcS0AKY6I+XeBTMDzQLX57XOH9Xg28pw='),
```
Also same :-D

elimisteve commented on index.js:
```diff
@@ -1,34 +1,82 @@
+      seedBytesUint8Array = new Uint8Array(scryptByteArray)
+
+      // First 32 Bytes of scrypt seed for encryption keys
+      // Note : first 32 Bytes are the same for dkLen 32 (old way) and 64!
```
This fact also seems to suggest that backward-compatibility was maintained, which is fantastic. /cc @jo @grempe

grempe commented:
Sorry for the delay in responding. I didn't test for compatibility with Minilock. But I think if the original version of this lib was compatible, then the PR I added also would be. I don't think the original version of this lib made any claims one way or the other about being compatible.

I think adding several tests that compare a key generated by minilock and session25519 would be good to add to this lib prior to relying on it and to ensure that any such compatibility doesn't break in the future.

elimisteve commented:
Hi @grempe, you're right; adding tests is important and I plan to add them as I get closer to relying on said compatibility. Thanks!
